docs(npx-version): mentioned using a renovate regex manager to update the version (#2786)
semantic-release/semantic-release: docs/usage/installation.md

$ npx semantic-release
```

### Notes

1. For a global installation, it's recommended to specify the major **semantic-release** version to install (for example with `npx semantic-release@18`).
   This way your build will not automatically use the next major **semantic-release** release that could possibly break your build.

>    You will have to upgrade manually when a new major version is released.

Author: i have not tested the regex manager config suggested in this PR against an npx command that specifies only the major version. it is unclear if it would update only the major version, replace the partial definition with a full one, or fail in some way. is there a difference in this case that is worth clarifying in some way, or should both scenarios be expected to work with the suggested config?

2. Specifying an exact version will make your releases even more deterministic, but requires regular maintenance to stay up to date when new versions of **semantic-release** become available.

>    In projects using [Renovate](https://docs.renovatebot.com/), a [regex manager](https://docs.renovatebot.com/modules/manager/regex/) can be defined to recommend version updates when new **semantic-release** versions are published:

Reviewer: ```suggestion
   You can use [Renovate's regex manager](https://docs.renovatebot.com/modules/manager/regex/) to get updates for **semantic-release**.
   Put this in your Renovate configuration file:
```

   ```json
   {
     "regexManagers": [
       {
         "description": "Update semantic-release version used by npx",
         "fileMatch": ["^\\.github/workflows/[^/]+\\.ya?ml$"],
         "matchStrings": ["\\srun: npx semantic-release@(?<currentValue>.*?)\\s"],
         "datasourceTemplate": "npm",
         "depNameTemplate": "semantic-release"
       }
     ]
   }
   ```

> 3. `npx` is a tool bundled with `npm@>=5.2.0`. It is used to conveniently install the semantic-release binary and to execute it.

Reviewer: ```suggestion
3. `npx` is a tool bundled with `npm@>=5.2.0`. You can use it to install (and run) the semantic-release binary.
```

$ npx semantic-release
```

### Notes

1. If you've globally installed **semantic-release** then we recommend that you set the major **semantic-release** version to install.
   For example, by using `npx semantic-release@18`.
   This way you control which major version of **semantic-release** is used by your build, and thus avoid breaking the build when there's a new major version of **semantic-release**.
   This also means you, or a bot, must upgrade **semantic-release** when a new major version is released.
2. Pinning **semantic-release** to an exact version makes your releases even more deterministic.
   But pinning also means you, or a bot, must update to newer versions of **semantic-release** more often.
   You can use [Renovate's regex manager](https://docs.renovatebot.com/modules/manager/regex/) to get updates for **semantic-release**.
   Put this in your Renovate configuration file:
   ```json
   {
     "regexManagers": [
       {
         "description": "Update semantic-release version used by npx",
         "fileMatch": ["^\\.github/workflows/[^/]+\\.ya?ml$"],

>          "matchStrings": ["\\srun: npx semantic-release@(?<currentValue>.*?)\\s"],

Reviewer: Is this matchStrings valid for both _pinned_ and _normal_ versions? Or just for one of those?

Author: good question. i wondered the similar. i tested with:
* [just the major version](https://github.com/travi-test/semantic-release-tester/pull/418/files)
* [a fully pinned version](https://github.com/travi-test/semantic-release-tester/pull/417/files)

both were recognized, since the expression ends with the whitespace after the version, and provided updates as i would have expected. i think this covers the scenarios suggested in this portion of the docs.

out of curiosity, i've also tried [specifying the version as simply `beta`](https://github.com/travi-test/semantic-release-tester/commit/d359f6051719a644270cae7542be59532702f06f) since i wasnt sure how that would be handled with no additional configuration. i did not receive an update PR, which is good since i normally use that when testing a pre-release channel while working through a series of updates. i'm still curious what would happen if i specified a precise pre-release version, like `semantic-release@22.0.0-beta.1`, but have not tried that out yet. i think that curiosity goes beyond the goals of this PR, though, so i can experiment further later.

Reviewer: > both were recognized, since the expression ends with the whitespace after the version, and provided updates as i would have expected. i think this covers the scenarios suggested in this portion of the docs.

Can you explain in the docs what works/doesn't work? That way readers don't have to guess.

> i'm still curious what would happen if i specified a precise pre-release version, like `semantic-release@22.0.0-beta.1`, but have not tried that out yet. i think that curiosity goes beyond the goals of this PR, though, so i can experiment further later.

I don't know if the `matchStrings` "version regex" is "wide" enough to include `-beta` versions. I don't know enough RegEx to check this. I think you should try it, and see what happens. 😄 

Also if you made it "wide" enough, so it captures `-beta` you should then find a way to avoid the `--` CLI flags.

Author: > Can you explain in the docs what works/doesn't work? That way readers don't have to guess.

i decided to split the mention of the regex manager to a separate note so that i could clarify that it was an automation option for either pinning strategy. i think that provides better clarity for those scenarios

> I don't know if the matchStrings "version regex" is "wide" enough to include -beta versions. I don't know enough RegEx to check this. I think you should try it, and see what happens. 😄

the `matchStrings` regex does match versions that include the pre-release version. i think those scenarios are beyond the scope of what these particular docs need to cover, so i'm comfortable investigating those details after getting these updates merged.

> Also if you made it "wide" enough, so it captures -beta you should then find a way to avoid the -- CLI flags.

since the regex only matches up to the next whitespace character, other trailing flags would already be excluded

Reviewer: Thanks for responding to me again `@travi`, and for improving the docs again.

I'll let you resolve this comment. 😉 I don't have the rights to do that, for whatever reason.

         "datasourceTemplate": "npm",
         "depNameTemplate": "semantic-release"
       }
     ]
   }
   ```

> 4. `npx` is a tool bundled with `npm@>=5.2.0`. You can use it to install (and run) the semantic-release binary.

Reviewer: I think **semantic-release** needs to be bold:

```suggestion
4. `npx` is a tool bundled with `npm@>=5.2.0`. You can use it to install (and run) the **semantic-release** binary.
```

Author: good catch
